### ISSUE TITLE
Better optimization for manually inlined code

### ISSUE THREAD

**Issue by MaxGraey** [WebAssembly/binaryen#1580]
Hi, @kripken!
In compilers like AssemblyScript which allow using manual inlining we not always get optimal optimization. 

Just let me take piratical example for better explanation of this issue in [this playground](https://webassembly.studio/?f=c1ge2xinw0w)

When we are inlining code with with some param constant propagation like that:
```ts
// Just some synthetic example
@inline // force inlining provide by asc but using binaryen for that
function shl(x: u64, shift: i32): u64 {
  // { block should be pre-calculated if shift is contant
  shift &= 127;
  var mod1: u64 = ((((shift + 127) | shift) & 64) >> 6) - 1;
  var mod2: u64 = (shift >> 6) - 1;
  shift &= 63;
  // }
  return (x & mod1) + ((x << 32) & mod2);
}

export function foo(a: u64): u64 {
  return shl(a, 1);
}
```
We expected something like this with `optimizeLevel = 3`:
```wast
(func $main/$foo (export "foo") (type $t0) (param $p0 i64) (result i64)
    get_local $p0
    i64.const 4294967295
    i64.and
    get_local $p0
    i64.const 32
    i64.shl
    i64.const 4294967295
    i64.and
    i64.add)
```
But currently we have codegen without constant simplification:
```wasm
(func $main/foo (export "foo") (type $t0) (param $p0 i64) (result i64)
    (local $l0 i32) (local $l1 i64)
    i32.const 1
    tee_local $l0
    i32.const 127
    i32.add
    i32.const 1
    i32.or
    i32.const 64
    i32.and
    i32.const 6
    i32.shr_s
    i32.const 1
    i32.sub
    i64.extend_u/i32
    set_local $l1
    get_local $p0
    i64.const 4294967295
    i64.and
    get_local $p0
    i64.const 32
    i64.shl
    i64.const 4294967295
    i64.and
    i64.add)
```

**Comment by kripken** [comment]
I'm not sure how to use that playground link - does it have the passes in there somehow? Can I run them there?

Overall, running the `optimizing-inlining` pass should do what you want. `inlining` just inlines, and you need to run opts yourself after, but `optimizing-inlining` will run function-level opts after inlining into a function, https://github.com/WebAssembly/binaryen/blob/master/src/passes/Inlining.cpp#L365 and those opts should perform constant simplification.

That pass is also run by default in `-O2` and above, https://github.com/WebAssembly/binaryen/blob/master/src/passes/pass.cpp#L190

If that doesn't work right for you, the simplest thing for me to debug would be a .wast file plus the binaryen passes to run on it (either in wasm-opt on the commandline, or binaryen.js commands).

**Comment by MaxGraey** [comment]
You can find some cli compiler flags In `gulpfile.js`.
```
    "main.ts",
    "--baseDir", "assembly",
    "--binaryFile", "../out/main.wasm",
    "--sourceMap",
    "--measure",
    "-O3"
```

ASC also support configure manual binaryen passes via:
```
   "--runPasses optimizing-inlining inlining"
```
And rebuild

**Comment by kripken** [comment]
I see, thanks.

So, first thing, it looks like AssemblyScript is doing the inlining. When I disable opts, I still see the inlining happen.

Second, running `--precompute-propagate -O3` does optimize it the way you want.

So basically, if binaryen did the inlining, it would have run precompute-propagate for you (see first link above). When not inlining, binaryen does run that pass, but late. Perhaps we should add it early as well, but the key information here, that we inlined, is not given to the binaryen optimizer. Therefore, since AssemblyScript does its own inlining, and so it knows about it, I'd recommend it also running precompute-propagate before running the main opts.

**Comment by MaxGraey** [comment]
Thanks, Alan! Addressed your explanation to @dcodeIO

**Comment by dcodeIO** [comment]
For a bit of background: What AssemblyScript does there is inline calls to functions annotated with an `@inline` decorator. That happens right away on code generation, compiling the function as a block, before any passes are being run. My expectation would be that this should optimize just fine like any other user provided code before optimizations, as the same could, theoretically, have been written by a programmer manually? The blocks are named `...|inlined` - might that cause trouble?

**Comment by kripken** [comment]
@dcodeIO Fair point, yeah, it could just have been written manually. My perspective was skewed because LLVM would have done it, so I didn't realize that, I guess...

Fix in #1581
